_posts/2021-04-26-integrating-airflow-and-okta.md

```diff
@@ -0,0 +1,96 @@
+1. In the “Allowed Grant Types” section, make sure you check all of the boxes.
+1. For the Login redirect URIs field, you will enter: `https://your-airflow-url-goes-here.com/oauth-authorized/okta`
+1. For the Initiate login URI field, you will enter: `https://your-airflow-url-goes-here.com/login`
+
+## Airflow Configuration
+
+`conf/webserver_config.py`
+
+    AUTH_TYPE = AUTH_OAUTH
+    OAUTH_PROVIDERS = [
+    {'name': 'okta', 'icon': 'fa-circle-o',
+            'token_key': 'access_token',
+            'remote_app': {
+                'client_id': <<>>,
+                'client_secret': <<>>,
+                'api_base_url': 'https://<<okta_url>>/oauth2/v1/',
+                'client_kwargs': {
+                    'scope': 'openid profile email groups'
+                },
+                'access_token_url': 'https://<<okta_url>>/oauth2/v1/token',
+                'authorize_url': 'https://<<okta_url>>/oauth2/v1/authorize',
+        }
+        }
+    ]
+
+### Special Steps:
+
+
+1. We started with Flask-AppBuilder 3.2.1, however it had a bug that needs to
+   be fixed, we raised a [PR](https://github.com/dpgaspar/Flask-AppBuilder/pull/1589) for Flask-AppBuilder:
+   fix: load user info for oktaCLOSED  to resolve that issue. That PR got
+   merged and now we can use the new release, Flask-AppBuilder 3.2.2
+
+2. As we were migrating from LDAP, we will already have user info populated,
+   however Okta generates a new user id something like
+   `thisokta_00u1046sqzJprt1hZ4x6`, but as the email id corresponding to that
+   user id is already present we got the below error. To prevent this we logged
+   into the underlying database for Airflow and cleaned up the ab_user and
+   ab_user_role table and let Okta integration recreate the user during first
+   sign up.
+
+    ```
+    [2021-03-19 16:32:28,559] {manager.py:215} ERROR - Error adding new user to database. (sqlite3.IntegrityError) UNIQUE constraint failed: ab_user.email
+    [SQL: INSERT INTO ab_user (first_name, last_name, username, password, active, email, last_login, login_count, fail_login_count, created_on, changed_on, created_by_fk, changed_by_fk) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]
+    [2021-03-19 16:32:28,560] {manager.py:1321} ERROR - Error creating a new OAuth user okta_00u1046sqzJprt1hZ4x6
+    ```
+3. Because we have deleted all the existing user and role, once the users logged in for the first time, especially for the first admin user we did the following from the airflow cli. This will create the first admin user after that if needed we can propagate other user and roles from the Airflow web console from this admin user account.
+    ```
+     airflow users add-role -r Admin -u okta_00u1046sqzJprt1hZ4x6
+    ```
+
+## Known Issue:
+
+1. Currently in the audit log, any action triggered on Airflow has Okta user id. Airflow needs to be patched to write out audit log entries with human readable user identifiers instead.
+
+---
+
```

Comment:
I think some form of narrative to wrap up the post, perhaps a screenshot of the Airflow tile in an Okta dashboard (with nothing else visible of course). "It's a little tricky to get this set up, but it's worth it!" or something like that :laughing: